src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -204,6 +205,39 @@ async ValueTask DecrementInFlightCountAsync(InFlightSolution inFlightSolution)
             }
         }
 
+        private static async Task<bool> IsIncrementalUpdateAsync(
+            AssetProvider assetProvider,
+            Checksum solutionChecksum,
+            Solution currentSolution,
+            CancellationToken cancellationToken)
+        {
+            var newSolutionCompilationChecksums = await assetProvider.GetAssetAsync<SolutionCompilationStateChecksums>(
+                assetHint: AssetHint.None, solutionChecksum, cancellationToken).ConfigureAwait(false);
+            var newSolutionChecksums = await assetProvider.GetAssetAsync<SolutionStateChecksums>(
+                assetHint: AssetHint.None, newSolutionCompilationChecksums.SolutionState, cancellationToken).ConfigureAwait(false);
+
+            var newSolutionInfo = await assetProvider.GetAssetAsync<SolutionInfo.SolutionAttributes>(
+                assetHint: AssetHint.None, newSolutionChecksums.Attributes, cancellationToken).ConfigureAwait(false);
+
+            // if either solution id or file path changed, then we consider it as new solution
+            return currentSolution.Id == newSolutionInfo.Id && currentSolution.FilePath == newSolutionInfo.FilePath;
+        }
+
+        private async Task<Solution> GetOrCreateSolutionToUpdateAsync(
+            AssetProvider assetProvider,
+            Checksum solutionChecksum,
+            CancellationToken cancellationToken)
+        {
+            // See if we can just incrementally update the current solution.
+            var currentSolution = this.CurrentSolution;
+            if (await IsIncrementalUpdateAsync(assetProvider, solutionChecksum, currentSolution, cancellationToken).ConfigureAwait(false))
+                return currentSolution;
+
+            // If not, have to create a new, fresh, solution instance to update.
+            var solutionInfo = await assetProvider.CreateSolutionInfoAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+            return CreateSolutionFromInfo(solutionInfo);
+        }
```

Comment:
simple logic.  determine if we're updating the existing solution (and return it if so), or creating a new solution (and create it if so).